include/snap/metric.h

```diff
@@ -12,35 +12,152 @@ See the License for the specific language governing permissions and
+
+namespace Plugin {
 
-// TODO (danielscottt): use a union for data here.
-class Metric {
+/**
+ * A PODS describing an element in a metric namespace.
+ */
+class Metric final {
 
   public:
-    void setData(int data);
+
+    struct NamespaceElement {
+      public:
+        /**
+         * value is the static value of this node in a namespace.
+         * When a namespace element is _not_ dynamic, value is used. During
+         * metric collection, value should contain the static name for this metric.
+         */
+        std::string value;
+
+        /**
+         * name is used to describe what this dynamic element is querying against.
+         * E.g. in the namespace `/intel/kvm/[vm_id]/cpu_wait` the element at index
+         * 2 has the name "vm_id".
+         * @see value
+         */
+        std::string name;
+
+        /**
+         * description is the description of this namespace element.
+         */
+        std::string description;
+    };
+
+    Metric();
+    /**
+     * The typical metric constructor.
+     * @param ns The metric's namespace.
+     * @param unit The metric's unit.
+     * @param description The metric's description.
+     */
+    Metric(std::vector<NamespaceElement> ns, std::string unit,
+           std::string description);
+
+    /**
+     * This constructor is used in the collector proxy during collect metrics.
+     * It's used to wrap the rpc::Metric type with the metric type from this
+     * library.
+     */
+   Metric(rpc::Metric* metric);
+
+    Metric(const Metric& from);
+
+    ~Metric();
+
+    /**
+     * dynamic_ns_elements returns the indices in the metric's namespace which
+     * are dynamic.
+     */
+    std::vector<int> dynamic_ns_elements();
+
+    /**
+     * ns returns the metric's namespace.
+     * If there is a memoized copy, that is returned. Else the namespace is
+     * copied into the cache then returned.
+     */
+    const std::vector<NamespaceElement>& ns();
+
+    /**
+     * set_ns sets the namespace of the metric in its `rpc::Metric` ptr.
+     * It also invalidates the memoization cache of the namespace if it is
+     * present.
+     * @see memo_ns
+     */
+    void set_ns(std::vector<NamespaceElement>);
+
+    /**
+     * tags returns the metric's tags.
+     * If there is a memoized copy, that is returned. Else the tags are copied
+     * into the cache then returned.
+     */
+    const std::map<std::string, std::string>& tags();
+
+    /**
+     * set_ns adds tags to the metric in its `rpc::Metric` ptr.
+     * It also invalidates the memoization cache of the tags if it is
+     * present.
+     * @see memo_tags_ptr
+     */
+    void add_tag(std::pair<std::string, std::string>);
+
+    /**
+     * set_timestamp sets the timestamp as now.
+     */
+    void set_timestamp();
+
+    /**
+     * set_timestamp sets the timestamp as tp.
+     * @param tp The timestamp to use.
+     */
+    void set_timestamp(std::chrono::system_clock::time_point tp);
```

Comment:
`std::chrono` was not fully supported in some compilers a few months back. Maybe worth considering making a list of supported compilers and C++ standards